packages/flutter/lib/src/rendering/sliver_group.dart

```diff
@@ -353,57 +353,38 @@ class RenderSliverMainAxisGroup extends RenderSliver
       hasVisualOverflow:
           totalScrollExtent > constraints.remainingPaintExtent || constraints.scrollOffset > 0.0,
     );
+
+    // Update the children's paintOffset based on the direction again, which
+    // must be done after obtaining the `paintExtent`.
```

Comment:
Ah I see! Well done! This is what tripped me up. The paintExtent here is like the viewport extent for normal sliver layout. Since we are laying out slivers inside of slivers, the paint extent is what we need to factor in. 💯